File: timeseriesjava/src/main/java/com/timeseries/StockSpan.java
import java.util.Stack;

package com.timeseries;


public class StockSpan {

    /**
     * Calculates the stock span values for the given array of prices.
     * The span of the stock's price on a given day is defined as the maximum number of consecutive days 
     * (starting from that day and going backwards) for which the price of the stock was less than or equal to its price on the given day.
     *
     * @param prices array of stock prices
     * @return array of span values
     */
    public static int[] calculateSpan(int[] prices) {
        int n = prices.length;
        int[] span = new int[n];
        Stack<Integer> stack = new Stack<>();

        for (int i = 0; i < n; i++) {
            while (!stack.isEmpty() && prices[stack.peek()] <= prices[i]) {
                stack.pop();
            }
            span[i] = stack.isEmpty() ? (i + 1) : (i - stack.peek());
            stack.push(i);
        }
        return span;
    }

    // Example usage
    public static void main(String[] args) {
        int[] prices = {100, 80, 60, 70, 60, 75, 85};
        int[] span = calculateSpan(prices);

        System.out.print("Stock Span: ");
        for (int s : span) {
            System.out.print(s + " ");
        }
    }
}

File: timeseriespy/stock_span.py
def calculate_stock_span(prices):
    """
    Calculate the stock span for each day.
    Span is the number of consecutive days before the current day
    where the price was less than or equal to today's price.

    Args:
        prices (list): List of stock prices.

    Returns:
        list: List of spans.
    """
    n = len(prices)
    span = [0] * n
    stack = []

    for i in range(n):
        # Pop prices from stack while stack is not empty and price at stack top is less than or equal to current price
        while stack and prices[stack[-1]] <= prices[i]:
            stack.pop()
        # If stack is empty, all previous prices are less than current price
        span[i] = i + 1 if not stack else i - stack[-1]
        stack.append(i)
    return span

# Example usage:
if __name__ == "__main__":
    prices = [100, 80, 60, 70, 60, 75, 85]
    print(calculate_stock_span(prices))  # Output: [1, 1, 1, 2, 1, 4, 6]